### ISSUE TITLE
Interaction with other context providers - how to go about it ?

### ISSUE THREAD

**Issue by ThisSentenceIsFalse** [ReactTraining/react-router#4335]
## **Prologue**

We're working on a web application resting on React and a (private) Flux-like store implementation. The team has been faring well with React Router v2 since the onset of the project, but I saw the latest changes in version 4 a few days ago and I'm seriously considering an upgrade.

As I've understood from the source code, `<Router/>` is a context provider passing `history` to at least `<Switch/>`,  `<Route/>` and `<Link>`. Our own components include context providers for store identifiers and other static metadata, and HoC's which use the implicit data. Some of these are interwoven, as in:

```
// provides some store id
<ReactForm id={}> 
  // arbitrary layout
  ...
  ...
  // adds field name to context
  <ReactForm.Field name={someName} {...otherProps}> 
    // complicated HTML: labels, field messages, icons etc...
    ...
    ...
    // connects to the store indicated by the store id 
    // **and** fetches field data as indicated by the field name
    <Field.Control type={Select} options={...} selection={...} {...rawInputProps}/>
    ...
    ...
  </ReactForm.Field>
  ...
  ...
</ReactForm>
```

I agree that this sort of context composing looks like it's asking for trouble - the advantage is that it makes complex layouts work out easily and readably. Other shortcomings I can think of are:

- given the way React allows access to context, we need some convention in order to make these components truly composable (by 'composable' I mean e.g. do not simply override outer context that you don't know about);
- never again shall I get a good sleep at night;

I also agree that the form example is contrived with respect to React Router, because I don't put routes in forms (links might be a different story). But as it is now, we would end up using elements of React Router along with the main layout, which does use the context feature.

# **TL;DR**

We fear that the `<Router/>` and other providers in my hierarchy might interfere with each other's context passing in the hierarchy - especially _in the main layout of the UI_ (i.e. wherein we render our different pages). This would complicate the interaction with the router.

To your knowledge, what approach would deal with this as painlessly as possible?

Would you consider allowing more control in the API for passing `history` around (e.g. exporting the raw versions of the components, which expect `history` as a prop, and somehow allowing access to the router history)?

## **Epilogue: What I have considered**

- wrapping `<Router/>` and all needed components to make them compatible with our conventions: not difficult, but unmaintainable now as there is no fixed API contract on the expected context;
- writing my own routing components around history: I don't want this.
- putting a `<Router/>` as the innermost provider, specifically for routes: I don't like this - as I understand it, it would result in one history for each point in the hierarchy where I need routing elements; I feel this would be a bad design decision;
- changing some of the premises for my components: long story short, it means a lot of work and I'm interested in keeping our work in the app as simple as possible;

I realise you have to consider many use cases and this particular architecture may not be of much interest. But the library you seem to be leaning to with v4 mixes routing with the rest of the hierarchy, hence the concern for how you may use context in your elements. Even if the end-user doesn't fiddle with that React features, it might be that other libraries are.

Thank you for the time invested in reading this, and sorry If I've horribly missed something.

**Comment by timdorr** [comment]
I'm not sure what the issue is here. What is the problem you're having? 

Taking a stab at it anyways: 4.0 and 2.x/3.0 also use context the same way. `this.context` is just a combination of whatever `contextTypes` requests from your ancestor components, so you can have multiple components providing it. We let you use it without having to concern yourself with the details via [`withRouter`](https://github.com/ReactTraining/react-router/blob/master/modules/withRouter.js) or [`withHistory` in 4.0](https://github.com/ReactTraining/react-router/blob/67c61692571c9ab236fb44df3b5916a5b77a0b02/modules/Core.js#L61-L87).
